File: apps/api/test_people_api.py
#!/usr/bin/env python3
"""
Test script for People API endpoints
"""

import requests
import json
from datetime import datetime

# API base URL
BASE_URL = "http://localhost:8000/api/v1"

# Test authentication credentials
TEST_USER = {
    "email": "test@arketic.com",
    "password": "Test123!@#"
}

def get_auth_token():
    """Get authentication token"""
    try:
        response = requests.post(f"{BASE_URL}/auth/login", json=TEST_USER)
        if response.status_code == 200:
            data = response.json()
            return data.get("access_token")
        else:
            print(f"Authentication failed: {response.status_code}")
            print(response.text)
            return None
    except Exception as e:
        print(f"Error during authentication: {e}")
        return None

def test_get_people(token):
    """Test GET /api/v1/organization/people endpoint"""
    headers = {"Authorization": f"Bearer {token}"}
    try:
        response = requests.get(f"{BASE_URL}/organization/people/", headers=headers)
        print(f"\nGET /organization/people/ - Status: {response.status_code}")
        if response.status_code == 200:
            data = response.json()
            print(f"Response structure: {list(data.keys())}")
            if "people" in data:
                print(f"Total people: {data.get('total', 0)}")
                print(f"Page: {data.get('page', 1)}")
                print(f"Page size: {data.get('page_size', 20)}")
                if data.get("people"):
                    print(f"\nFirst person:")
                    print(json.dumps(data["people"][0], indent=2))
            return True
        else:
            print(f"Error: {response.text}")
            return False
    except Exception as e:
        print(f"Error: {e}")
        return False

def test_create_person(token):
    """Test POST /api/v1/organization/people endpoint"""
    headers = {"Authorization": f"Bearer {token}"}
    
    test_person = {
        "first_name": "Test",
        "last_name": "User",
        "email": f"testuser_{datetime.now().strftime('%Y%m%d%H%M%S')}@arketic.com",
        "phone": "+1-555-0123",
        "job_title": "Software Developer",
        "department": "Engineering",
        "site": "Main Office",
        "location": "New York, NY",
        "role": "User",
        "notes": "Test user created via API test script"
    }
    
    try:
        response = requests.post(f"{BASE_URL}/organization/people/", headers=headers, json=test_person)
        print(f"\nPOST /organization/people/ - Status: {response.status_code}")
        if response.status_code == 201:
            data = response.json()
            print(f"Created person successfully:")
            print(json.dumps(data, indent=2))
            return data.get("id")
        else:
            print(f"Error: {response.text}")
            return None
    except Exception as e:
        print(f"Error: {e}")
        return None

def test_update_person(token, person_id):
    """Test PUT /api/v1/organization/people/{id} endpoint"""
    headers = {"Authorization": f"Bearer {token}"}
    
    update_data = {
        "job_title": "Senior Software Developer",
        "notes": "Updated via API test script"
    }
    
    try:
        response = requests.put(f"{BASE_URL}/organization/people/{person_id}", headers=headers, json=update_data)
        print(f"\nPUT /organization/people/{person_id} - Status: {response.status_code}")
        if response.status_code == 200:
            data = response.json()
            print(f"Updated person successfully:")
            print(f"  Job Title: {data.get('job_title')}")
            print(f"  Notes: {data.get('notes')}")
            return True
        else:
            print(f"Error: {response.text}")
            return False
    except Exception as e:
        print(f"Error: {e}")
        return False

def test_delete_person(token, person_id):
    """Test DELETE /api/v1/organization/people/{id} endpoint"""
    headers = {"Authorization": f"Bearer {token}"}
    
    try:
        response = requests.delete(f"{BASE_URL}/organization/people/{person_id}", headers=headers)
        print(f"\nDELETE /organization/people/{person_id} - Status: {response.status_code}")
        if response.status_code == 204:
            print(f"Deleted person successfully")
            return True
        else:
            print(f"Error: {response.text}")
            return False
    except Exception as e:
        print(f"Error: {e}")
        return False

def main():
    print("=" * 60)
    print("Testing People API Endpoints")
    print("=" * 60)
    
    # Get authentication token
    print("\n1. Getting authentication token...")
    token = get_auth_token()
    if not token:
        print("Failed to authenticate. Exiting.")
        return
    
    print(f"   Token obtained: {token[:20]}...")
    
    # Test GET people
    print("\n2. Testing GET /organization/people/...")
    test_get_people(token)
    
    # Test CREATE person
    print("\n3. Testing POST /organization/people/...")
    person_id = test_create_person(token)
    
    if person_id:
        # Test UPDATE person
        print("\n4. Testing PUT /organization/people/{id}...")
        test_update_person(token, person_id)
        
        # Test DELETE person
        print("\n5. Testing DELETE /organization/people/{id}...")
        test_delete_person(token, person_id)
    
    print("\n" + "=" * 60)
    print("All tests completed!")
    print("=" * 60)

if __name__ == "__main__":
    main()

File: docker-auth-tests/test-people-api.js
#!/usr/bin/env node

/**
 * People API Test Suite
 * Tests the People management endpoints
 */

const axios = require('axios');

const API_BASE = 'http://localhost:8000';

async function testPeopleAPI() {
    console.log('🧪 Testing People API...\n');
    
    try {
        // Step 1: Login to get auth token
        console.log('Step 1: Login...');
        const loginResponse = await axios.post(`${API_BASE}/api/v1/auth/login`, {
            username: "admin@arketic.com",
            password: "admin123"
        });
        
        if (!loginResponse.data.access_token) {
            throw new Error('Login failed - no access token');
        }
        
        const token = loginResponse.data.access_token;
        const headers = { Authorization: `Bearer ${token}` };
        
        console.log('✅ Login successful');
        
        // Step 2: Create test person
        console.log('\nStep 2: Create Person...');
        const personData = {
            first_name: "John",
            last_name: "Doe", 
            email: "john.doe@company.com",
            phone: "+1-555-123-4567",
            job_title: "Software Engineer",
            department: "Engineering",
            site: "Head Office",
            role: "User",
            hire_date: "2024-01-15T00:00:00Z",
            manager_id: null,
            location: "San Francisco",
            employee_id: "EMP001"
        };
        
        const createResponse = await axios.post(
            `${API_BASE}/api/v1/organization/people/`,
            personData,
            { headers }
        );
        
        if (createResponse.status !== 201) {
            throw new Error(`Create person failed with status: ${createResponse.status}`);
        }
        
        const createdPerson = createResponse.data;
        console.log('✅ Person created:', createdPerson.full_name);
        console.log(`   ID: ${createdPerson.id}`);
        console.log(`   Email: ${createdPerson.email}`);
        
        // Step 3: Get person by ID
        console.log('\nStep 3: Get Person by ID...');
        const getResponse = await axios.get(
            `${API_BASE}/api/v1/organization/people/${createdPerson.id}`,
            { headers }
        );
        
        console.log('✅ Person retrieved:', getResponse.data.full_name);
        
        // Step 4: List people
        console.log('\nStep 4: List People...');
        const listResponse = await axios.get(
            `${API_BASE}/api/v1/organization/people/`,
            { headers }
        );
        
        console.log('✅ People list retrieved:');
        console.log(`   Total: ${listResponse.data.total}`);
        console.log(`   Page: ${listResponse.data.page}`);
        console.log(`   Page size: ${listResponse.data.page_size}`);
        
        // Step 5: Update person
        console.log('\nStep 5: Update Person...');
        const updateData = {
            job_title: "Senior Software Engineer",
            department: "Engineering",
            phone: "+1-555-123-9999"
        };
        
        const updateResponse = await axios.put(
            `${API_BASE}/api/v1/organization/people/${createdPerson.id}`,
            updateData,
            { headers }
        );
        
        console.log('✅ Person updated:', updateResponse.data.job_title);
        
        // Step 6: Search people
        console.log('\nStep 6: Search People...');
        const searchResponse = await axios.get(
            `${API_BASE}/api/v1/organization/people/?query=John&department=Engineering`,
            { headers }
        );
        
        console.log('✅ Search completed:', searchResponse.data.people.length, 'results');
        
        // Step 7: Get stats
        console.log('\nStep 7: Get People Stats...');
        const statsResponse = await axios.get(
            `${API_BASE}/api/v1/organization/people/stats/overview`,
            { headers }
        );
        
        console.log('✅ Stats retrieved:');
        console.log(`   Total people: ${statsResponse.data.total_people}`);
        console.log(`   Active people: ${statsResponse.data.active_people}`);
        console.log(`   Departments:`, statsResponse.data.departments);
        
        // Step 8: Get hierarchy
        console.log('\nStep 8: Get Organization Hierarchy...');
        const hierarchyResponse = await axios.get(
            `${API_BASE}/api/v1/organization/people/hierarchy`,
            { headers }
        );
        
        console.log('✅ Hierarchy retrieved:');
        console.log(`   Total people in hierarchy: ${hierarchyResponse.data.total_people}`);
        
        // Step 9: Create another person as manager
        console.log('\nStep 9: Create Manager Person...');
        const managerData = {
            first_name: "Jane",
            last_name: "Smith", 
            email: "jane.smith@company.com",
            job_title: "Engineering Manager",
            department: "Engineering",
            role: "Manager",
            employee_id: "EMP002"
        };
        
        const createManagerResponse = await axios.post(
            `${API_BASE}/api/v1/organization/people/`,
            managerData,
            { headers }
        );
        
        const createdManager = createManagerResponse.data;
        console.log('✅ Manager created:', createdManager.full_name);
        
        // Step 10: Update first person to have the manager
        console.log('\nStep 10: Assign Manager...');
        const assignManagerResponse = await axios.put(
            `${API_BASE}/api/v1/organization/people/${createdPerson.id}`,
            { manager_id: createdManager.id },
            { headers }
        );
        
        console.log('✅ Manager assigned successfully');
        
        // Step 11: Test bulk operations
        console.log('\nStep 11: Test Bulk Create...');
        const bulkData = {
            people: [
                {
                    first_name: "Alice",
                    last_name: "Johnson",
                    email: "alice.johnson@company.com",
                    department: "Marketing",
                    role: "User"
                },
                {
                    first_name: "Bob",
                    last_name: "Wilson",
                    email: "bob.wilson@company.com", 
                    department: "Sales",
                    role: "User"
                }
            ]
        };
        
        const bulkResponse = await axios.post(
            `${API_BASE}/api/v1/organization/people/bulk`,
            bulkData,
            { headers }
        );
        
        console.log('✅ Bulk create successful:', bulkResponse.data.length, 'people created');
        
        // Step 12: Final people count
        console.log('\nStep 12: Final People Count...');
        const finalListResponse = await axios.get(
            `${API_BASE}/api/v1/organization/people/`,
            { headers }
        );
        
        console.log('✅ Final people count:', finalListResponse.data.total);
        
        console.log('\n🎉 All People API tests passed!\n');
        
        return {
            success: true,
            totalPeople: finalListResponse.data.total,
            createdPeople: [createdPerson, createdManager, ...bulkResponse.data]
        };
        
    } catch (error) {
        console.error('❌ People API test failed:', error.response?.data || error.message);
        
        if (error.response) {
            console.error('Response status:', error.response.status);
            console.error('Response headers:', error.response.headers);
        }
        
        return {
            success: false,
            error: error.message
        };
    }
}

// Run test
if (require.main === module) {
    testPeopleAPI().then(result => {
        console.log('Test Result:', result);
        process.exit(result.success ? 0 : 1);
    });
}

module.exports = { testPeopleAPI };